processing/src/main/java/org/apache/druid/segment/join/JoinFilterAnalyzer.java

```diff
@@ -0,0 +1,739 @@
+  }
+
+  /**
+   * Potentially rewrite the subfilters of an OR filter so that the whole OR filter can be pushed down to
+   * the base table.
+   *
+   * @param adapter          Adapter for the join
+   * @param orFilter         OrFilter to be rewritten
+   * @param prefixes         Map of table prefixes to clauses
+   * @param equiconditions   Map of equiconditions
+   * @param correlationCache Column correlation analysis cache
+   *
+   * @return A JoinFilterAnalysis indicating how to handle the potentially rewritten filter
+   */
+  @Nullable
+  private static JoinFilterAnalysis rewriteOrFilter(
+      HashJoinSegmentStorageAdapter adapter,
+      OrFilter orFilter,
+      Map<String, JoinableClause> prefixes,
+      Map<String, Expr> equiconditions,
+      Map<String, List<JoinFilterColumnCorrelationAnalysis>> correlationCache
+  )
+  {
+    boolean retainRhs = false;
+
+    List<Filter> newFilters = new ArrayList<>();
+    for (Filter filter : orFilter.getFilters()) {
+      boolean allBaseColumns = true;
+      for (String requiredColumn : filter.getRequiredColumns()) {
+        if (!adapter.isBaseColumn(requiredColumn)) {
+          allBaseColumns = false;
+        }
+      }
+
+      if (!allBaseColumns) {
+        retainRhs = true;
+        if (filter instanceof SelectorFilter) {
+          JoinFilterAnalysis rewritten = rewriteSelectorFilter(
+              adapter,
+              filter,
+              prefixes,
+              equiconditions,
+              correlationCache
+          );
+          if (!rewritten.isCanPushDown()) {
+            return JoinFilterAnalysis.createNoPushdownFilterAnalysis(orFilter);
+          } else {
+            newFilters.add(rewritten.getPushdownFilter());
+          }
+        } else {
+          return JoinFilterAnalysis.createNoPushdownFilterAnalysis(orFilter);
+        }
+      } else {
+        newFilters.add(filter);
+      }
+    }
+
+    return new JoinFilterAnalysis(
+        true,
+        retainRhs,
+        orFilter,
+        new OrFilter(newFilters),
+        null
+    );
+  }
+
+  /**
+   * Rewrites a selector filter on a join table into an IN filter on the base table.
+   *
+   * @param baseAdapter      The adapter for the join
+   * @param filter           Filter to be rewritten
+   * @param prefixes         Map of join table prefixes to clauses
+   * @param equiconditions   Map of equiconditions
+   * @param correlationCache Cache of column correlation analyses
+   *
+   * @return A JoinFilterAnalysis that indicates how to handle the potentially rewritten filter
+   */
+  private static JoinFilterAnalysis rewriteSelectorFilter(
+      HashJoinSegmentStorageAdapter baseAdapter,
+      Filter filter,
+      Map<String, JoinableClause> prefixes,
+      Map<String, Expr> equiconditions,
+      Map<String, List<JoinFilterColumnCorrelationAnalysis>> correlationCache
+  )
+  {
+    assert (filter instanceof SelectorFilter);
+    SelectorFilter selectorFilter = (SelectorFilter) filter;
```

Comment:
Can `rewriteSelectorFilter`'s `filter` parameter be of type `SelectorFilter` instead?

Reply:
Changed to SelectorFilter